crates/ruff_python_formatter/src/lib.rs

```diff
@@ -23,22 +29,52 @@ mod trivia;
 
 include!("../../ruff_formatter/shared_traits.rs");
 
-pub fn fmt(contents: &str) -> Result<Formatted<ASTFormatContext>> {
-    // Create a reusable locator.
-    let locator = Locator::new(contents);
+pub fn fmt(contents: &str) -> Result<Printed> {
+    // Tokenize once
+    let mut tokens = Vec::new();
+    let mut comment_ranges = CommentRangesBuilder::default();
 
-    // Tokenize once.
-    let tokens: Vec<LexResult> = ruff_rustpython::tokenize(contents);
+    for result in lex(contents, Mode::Module) {
+        let (token, range) = match result {
+            Ok((token, range)) => (token, range),
+            Err(err) => return Err(anyhow!("Source contains syntax errors {err:?}")),
+        };
+
+        comment_ranges.visit_token(&token, range);
+        tokens.push(Ok((token, range)));
+    }
+
+    let comment_ranges = comment_ranges.finish();
 
-    // Extract trivia.
     let trivia = trivia::extract_trivia_tokens(&tokens);
 
     // Parse the AST.
-    let python_ast = ruff_rustpython::parse_program_tokens(tokens, "<filename>")?;
+    let python_ast = parse_tokens(tokens, Mode::Module, "<filename>").unwrap();
+
+    let formatted = format_node(&python_ast, &comment_ranges, contents, trivia)?;
+
+    formatted
+        .print()
+        .with_context(|| "Failed to print the formatter IR")
+}
+
+pub(crate) fn format_node<'a>(
+    root: &'a Mod,
+    comment_ranges: &'a CommentRanges,
+    source: &'a str,
+    trivia: Vec<TriviaToken>,
+) -> FormatResult<Formatted<ASTFormatContext<'a>>> {
+    let comments = Comments::from_ast(root, SourceCode::new(source), comment_ranges);
+
+    let module = root.as_module().unwrap();
+
+    let locator = Locator::new(source);
 
     // Convert to a CST.
-    let mut python_cst: Vec<Stmt> = python_ast
-        .into_iter()
+    let mut python_cst: Vec<Stmt> = module
+        .body
+        .iter()
+        .cloned()
```

Comment:
This is obviously bad but I didn't bother to spend any time on fixing it because I want to remove the CST. I need to think about how to best transition from old to new.

Reply:
Remaining calm...